File: ndl.go
package main

import (
	"fmt"
	"os"
	"time"

	"environment"
	"haystack"
	"pipeline"
	"search"
)

func main() {
	start := time.Now()

	// construct the user's environment (includes parsing command line)
	env := environment.New()
	if env == nil {
		return
	}
	query := env.Query

	// maybe punt to grep formatting mode
	if query == "--reformat-grep-output" {
		mainReformatGrepOutput(env)
		return
	}

	// convert user's needle into a search
	s := search.New(query)

	// convert user's request into a haystack
	h := haystack.New(env, query)

	// make a pipeline to generate output
	p := pipeline.New(env, h, s)
	err := p.Run()
	if err != nil {
		panic(err)
	}

	msg := fmt.Sprintf("Runtime: %s", time.Now().Sub(start))
	env.WriteHeader(os.Stderr, msg)
}

File: ndl.pl
:- ['src/environment.pl'].
:- ['src/pipeline.pl'].
:- ['src/reformat-grep-output.pl'].

main(['--reformat-grep-output',Pattern]) :-
    !,
    reformat_grep_output(Pattern).
main(Args) :-
    get_time(Start),
    assert_environment(Args),

    phrase(pipeline,Pipeline),
    warn("~s~n", [Pipeline]),

    shell(Pipeline, Status),
    get_time(Stop),
    runtime(Start, Stop),
    halt(Status).


runtime(Start,Stop) :-
    Duration is 1000*(Stop - Start),
    format("Runtime: ~1f ms~n", [Duration]).


warn(Format,Args) :-
    format(user_error,Format,Args).
